### PR TITLE
#1707 fix alpha beta load disconnection and add log to timeline

### LINKED ISSUE #1649: Add log when state change in load model

### LINKED ISSUE #1707: Fix state evaluation for restorative alpha beta cpp model

### PR DESCRIPTION
**Pull request by marcochiaramello**
Closes #1707 
Closes #1649

### CODE REVIEW

File: dynawo/sources/Models/CPP/Components/ModelLoadRestorativeWithLimits/test/TestLoadRestorativeWithLimits.cpp [hunk 1]
```diff
@@ -263,7 +267,7 @@ TEST(ModelsLoadRestorativeWithLimits, ModelLoadRestorativeWithLimitsContinuousAn
   y[1] = 0;
   y[2] = 0;
   y[3] = 0;
-  ASSERT_THROW_DYNAWO(modelLoad->evalF(0, ALGEBRAIC_EQ), Error::NUMERICAL_ERROR, KeyError_t::NumericalErrorFunction);
+  // ASSERT_THROW_DYNAWO(modelLoad->evalF(0, ALGEBRAIC_EQ), Error::NUMERICAL_ERROR, KeyError_t::NumericalErrorFunction);
```
Comment (cossartque) [inline dynawo/sources/Models/CPP/Components/ModelLoadRestorativeWithLimits/test/TestLoadRestorativeWithLimits.cpp:270]: Shall we remove this line as it is a comment now ?

Comment (marcochiaramello) [reply dynawo/sources/Models/CPP/Components/ModelLoadRestorativeWithLimits/test/TestLoadRestorativeWithLimits.cpp:270]: yes, forgot it, thanks!

**Review by cossartque — APPROVED** [review]

**Comment by gautierbureau** [comment]
I would suggest to harmonize the handle of connection/disconnection with other private load models (or even other models that I don't think of). @adriguir you could also comment. thanks

**Comment by marcochiaramello** [comment]
I agree on the alignment, it stays to decide which version is more readable :)

**Comment by adriguir** [comment]
> I would suggest to harmonize the handle of connection/disconnection with other private load models (or even other models that I don't think of). @adriguir you could also comment. thanks

I always think harmonization should be a target! 
Then it depends on what we call harmonization and what you had in mind behind your comment.

From my point of view, what I prefer in the private repo implementation and what will motivate my recommendation to adopt it, is the fact that we raise a root change based on the discrete variables of the models themselves. In the implementation adopted in the LoadAlphaBetaRestorative... (I missed this point in the model review), there is no root change associated to the change in z which can led to misbehavior with some sequences and with some potential optimizations in the z propagation effect loops. 
I find it safer to be sure to raise the root if the z is evolving because it will evolve based on a connection in some of the cases (direct load disconnection) and in this case, z in the load would have changed by copy but if it doesn't impact the roots, I think we don't evaluate again the z variables. To sum up, I'm not completely sure that load disconnection works fine in every case with the current implementation.

**Comment by marcochiaramello** [comment]
> > I would suggest to harmonize the handle of connection/disconnection with other private load models (or even other models that I don't think of). @adriguir you could also comment. thanks
> 
> I always think harmonization should be a target!
> Then it depends on what we call harmonization and what you had in mind behind your comment.
> 
> From my point of view, what I prefer in the private repo implementation and what will motivate my recommendation to adopt it, is the fact that we raise a root change based on the discrete variables of the models themselves. In the implementation adopted in the LoadAlphaBetaRestorative... (I missed this point in the model review), there is no root change associated to the change in z which can led to misbehavior with some sequences and with some potential optimizations in the z propagation effect loops.
> I find it safer to be sure to raise the root if the z is evolving because it will evolve based on a connection in some of the cases (direct load disconnection) and in this case, z in the load would have changed by copy but if it doesn't impact the roots, I think we don't evaluate again the z variables. To sum up, I'm not completely sure that load disconnection works fine in every case with the current implementation.

Hello again,
fine for me, we just need to keep in mind that we will have then a gLocal variable more. I do not know if there is a way to get rid of it ...

File: dynawo/sources/Models/CPP/Components/ModelLoadRestorativeWithLimits/DYNModelLoadRestorativeWithLimits.cpp [hunk 2]
```diff
@@ -73,6 +73,8 @@ namespace DYN {
   UiYNum_(0),
   IrYNum_(0),
   IiYNum_(0),
+  preSwitchOff1_(0),
+  preSwitchOff2_(0),
```
Comment (rosiereflo) [inline dynawo/sources/Models/CPP/Components/ModelLoadRestorativeWithLimits/DYNModelLoadRestorativeWithLimits.cpp:77]: This should be initialized similarly to zLocal (with false)

File: dynawo/sources/Models/CPP/Components/ModelLoadRestorativeWithLimits/test/TestLoadRestorativeWithLimits.cpp [hunk 3]
```diff
@@ -259,11 +263,6 @@ TEST(ModelsLoadRestorativeWithLimits, ModelLoadRestorativeWithLimitsContinuousAn
   ASSERT_NO_THROW(modelLoad->evalMode(0));
   ASSERT_NO_THROW(modelLoad->setGequations());
   ASSERT_NO_THROW(modelLoad->evalF(1, ALGEBRAIC_EQ));
-
-  y[1] = 0;
```
Comment (rosiereflo) [inline dynawo/sources/Models/CPP/Components/ModelLoadRestorativeWithLimits/test/TestLoadRestorativeWithLimits.cpp:263]: why do we remove this?

**Review by rosiereflo — CHANGES_REQUESTED** [review]

File: dynawo/sources/Models/CPP/Components/ModelLoadRestorativeWithLimits/DYNModelLoadRestorativeWithLimits.cpp (hunk 2 above)
Comment (marcochiaramello) [reply dynawo/sources/Models/CPP/Components/ModelLoadRestorativeWithLimits/DYNModelLoadRestorativeWithLimits.cpp:77]: implementation has changed to align to other cpp models, those attributes are not anymore present

File: dynawo/sources/Models/CPP/Components/ModelLoadRestorativeWithLimits/test/TestLoadRestorativeWithLimits.cpp (hunk 3 above)
Comment (marcochiaramello) [reply dynawo/sources/Models/CPP/Components/ModelLoadRestorativeWithLimits/test/TestLoadRestorativeWithLimits.cpp:263]: it is a wrong delete I did, it is back there

File: dynawo/sources/Models/CPP/Components/ModelLoadRestorativeWithLimits/DYNModelLoadRestorativeWithLimits.cpp [hunk 4]
```diff
@@ -364,19 +380,23 @@ namespace DYN {
...
       case QNum_:
-        if (isConnected()) {
+        if (static_cast<int>(zLocal_[running]) == RUNNING_TRUE) {
           double U = sqrt(yLocal_[UrYNum_] * yLocal_[UrYNum_] + yLocal_[UiYNum_] * yLocal_[UiYNum_]);
           output = Q0Pu_ * pow_dynawo(U/yLocal_[UfYNum_], beta_);
         }
         break;
       case loadStateNum_:
-        output = connectionState_;
+      if (static_cast<int>(zLocal_[running]) == RUNNING_TRUE) {
```
Comment (rosiereflo) [inline dynawo/sources/Models/CPP/Components/ModelLoadRestorativeWithLimits/DYNModelLoadRestorativeWithLimits.cpp:395]: indentation

File: dynawo/sources/Models/CPP/Components/ModelLoadRestorativeWithLimits/DYNModelLoadRestorativeWithLimits.cpp [hunk 5]
```diff
@@ -430,14 +451,14 @@ namespace DYN {
...
-        if (isConnected()) {
+        if (static_cast<int>(zLocal_[running]) == RUNNING_TRUE) {
           numVars.push_back(UrYNum_);
           numVars.push_back(UiYNum_);
           numVars.push_back(UfYNum_);
         }
         break;
       case QNum_:
-        if (isConnected()) {
+        if (static_cast<int>(zLocal_[running]) == RUNNING_TRUE) {
```
Comment (rosiereflo) [inline dynawo/sources/Models/CPP/Components/ModelLoadRestorativeWithLimits/DYNModelLoadRestorativeWithLimits.cpp:461]: maybe we could keep isConnected() and replace its content with this static cast?

**Review by rosiereflo — CHANGES_REQUESTED** [review]

File: dynawo/sources/Models/CPP/Components/ModelLoadRestorativeWithLimits/DYNModelLoadRestorativeWithLimits.cpp (hunk 4 above)
Comment (marcochiaramello) [reply dynawo/sources/Models/CPP/Components/ModelLoadRestorativeWithLimits/DYNModelLoadRestorativeWithLimits.cpp:395]: ok, thanks

**Review by cossartque — APPROVED** [review]

**Review by gautierbureau — APPROVED** [review]

**Review by rosiereflo — APPROVED** [review]

**Pull request merged**